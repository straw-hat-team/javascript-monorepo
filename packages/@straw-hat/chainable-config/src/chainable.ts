import { Configurable } from './configurable';

export type BatchHandler = (context: Chainable<any>) => any;

export class Chainable<P> extends Configurable {
  protected parent: P;

  constructor(parent: P) {
    super();
    this.parent = parent;
  }

  static isChainable<P = unknown>(value: unknown): value is Chainable<P> {
    return value instanceof Chainable;
  }

  batch(handler: BatchHandler) {
    handler(this);
    return this;
  }

  end() {
    return this.parent;
  }

  when(
    predicate: boolean,
    whenTruthy: BatchHandler = Function.prototype as BatchHandler,
    whenFalsy: BatchHandler = Function.prototype as BatchHandler,
  ) {
    const callback = predicate ? whenTruthy : whenFalsy;
    callback(this);
    return this;
  }
}
