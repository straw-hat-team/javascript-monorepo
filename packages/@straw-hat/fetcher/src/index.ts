export * from './errors';
export * from './fetcher';
export * from './helpers';
export { composeMiddleware, createMiddleware } from './middlewares/middleware';
export * from './request';
export { cloneRequest } from './request';
