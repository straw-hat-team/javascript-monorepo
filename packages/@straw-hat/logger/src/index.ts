export * from './levels';
export * from './log-event';
export * from './log-handler';
export * from './logger';
export * from './noop-logger';
export * from './types';
