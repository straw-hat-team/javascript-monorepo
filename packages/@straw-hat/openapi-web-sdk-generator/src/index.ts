export * from './openapi-web-sdk-generator.js';
