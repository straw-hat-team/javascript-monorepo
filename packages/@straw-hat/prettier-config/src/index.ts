import type { Options } from 'prettier';
import baseConfig from './base.js';

export default {
  ...baseConfig,
  organizeImportsSkipDestructiveCodeActions: true,
  plugins: [...baseConfig.plugins, import.meta.resolve('prettier-plugin-organize-imports')],
} satisfies Options;
