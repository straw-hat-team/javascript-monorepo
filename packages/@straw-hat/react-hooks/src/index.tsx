export { createContextHook } from './create-context-hook';
export { useBoolean } from './use-boolean';
export { useCallOnce } from './use-call-once';
export { useControlled } from './use-controlled';
export { useEventListener } from './use-event-listener';
export { useIsomorphicLayoutEffect } from './use-isomorphic-layout-effect';
export { useLatest } from './use-latest';
export { usePrevious } from './use-previous';
export { useScript } from './use-script';
