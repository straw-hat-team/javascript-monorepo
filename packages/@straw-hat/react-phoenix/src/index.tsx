import { Channel, Socket, SocketConnectOption } from 'phoenix';
import React, { createContext, useContext, useEffect, useRef, useState } from 'react';

type SocketState = { socket: Socket | null };

export const SocketContext = createContext<SocketState | undefined>(undefined);

function useSocket(props: { url: string; options?: Partial<SocketConnectOption> }) {
  const [socket, setSocket] = useState<Socket | null>(null);

  useEffect(() => {
    const s = new Socket(props.url, props.options);
    s.connect();
    setSocket(s);

    return () => {
      s.disconnect();
      setSocket(null);
    };
  }, [props.options, props.url]);

  return {
    socket,
  };
}

export function SocketProvider(props: {
  children: React.ReactNode;
  options?: Partial<SocketConnectOption>;
  url: string;
}) {
  const value = useSocket({
    options: props.options,
    url: props.url,
  });
  return <SocketContext.Provider value={value}>{props.children}</SocketContext.Provider>;
}

function useSocketContext() {
  const value = useContext(SocketContext);
  if (value === undefined) {
    throw new Error('useSocket must be used within a SocketProvider');
  }
  return value.socket;
}

export function useChannel<Message extends unknown>(props: {
  enabled?: boolean;
  topic: string;
  params?: object;
  onJoin?: (channel: Channel, message: Message) => void;
  onError?: (channel: Channel, message: Message) => void;
  onTimeout?: () => void;
}) {
  const { enabled = true } = props;

  const socket = useSocketContext();
  const [channel, setChannel] = useState<Channel | null>(null);
  const onJoin = useRef(props.onJoin);
  const onError = useRef(props.onError);
  const onTimeout = useRef(props.onTimeout);

  useEffect(() => {
    if (socket === null || !enabled) {
      return;
    }

    const ch = socket.channel(props.topic, props.params);

    ch.join()
      .receive('ok', (message) => onJoin.current?.(ch, message))
      .receive('error', (message) => onError.current?.(ch, message))
      .receive('timeout', () => onTimeout.current?.());

    setChannel(ch);
    return () => {
      ch.leave();
      setChannel(null);
    };
  }, [socket, props.topic, props.params, enabled]);

  return channel;
}

export function useChannelEvent<Message extends unknown>(props: {
  channel: Channel | null;
  event: string;
  onMessage: (message: Message) => void;
}) {
  const onMessage = useRef(props.onMessage);

  useEffect(() => {
    const ref = props.channel?.on(props.event, onMessage.current);

    return () => {
      props.channel?.off(props.event, ref);
    };
  }, [props.channel, props.event]);
}
