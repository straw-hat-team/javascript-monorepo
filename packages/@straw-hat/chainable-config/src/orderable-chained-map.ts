import { Chainable } from './chainable';
import { ChainedMap } from './chained-map';

export enum OrderPositions {
  Before = 'before',
  After = 'after',
}

export type MoveCallback = (args: { before(relativeKey: string): void; after(relativeKey: string): void }) => any;

function relativePosition(index: number, order: OrderPositions) {
  return order === OrderPositions.Before ? index : index + 1;
}

function byKey(key: string) {
  return (entry: [string, any]) => entry[0] === key;
}

type MoveableValueArgs<T> = {
  parent: OrderableChainedMap<any, any>;
  key: string;
  value: T;
};

class MoveableValue<P, T> extends Chainable<OrderableChainedMap<P>> {
  private value: T;
  private key: string;

  constructor(args: MoveableValueArgs<T>) {
    super(args.parent);
    this.key = args.key;
    this.value = args.value;
  }

  override valueOf() {
    return this.value;
  }

  override toConfig() {
    return this.value;
  }

  before(key: string) {
    this.parent.move(this.key, ({ before }) => before(key));
    return this;
  }

  after(key: string) {
    this.parent.move(this.key, ({ after }) => after(key));
    return this;
  }
}

export class OrderableChainedMap<P, S = unknown> extends ChainedMap<P, MoveableValue<P, S>> {
  // @ts-ignore TODO: figure out how to fix this issue
  set<T extends S>(key: string, value: T, options: { alias?: string } = {}) {
    const theKey = options.alias ?? key;

    const item = new MoveableValue<P, T>({
      parent: this,
      key: theKey,
      value,
    });

    super.set(key, item, options);
    return this;
  }

  move(key: string, callback: MoveCallback) {
    callback({
      before: (relativeKey: string) => {
        this.#doMove(key, OrderPositions.Before, relativeKey);
      },
      after: (relativeKey: string) => {
        this.#doMove(key, OrderPositions.After, relativeKey);
      },
    });

    return this;
  }

  #doMove = (key: string, order: OrderPositions, relativeKey: string) => {
    if (!this.store.has(key) || !this.store.has(relativeKey)) {
      return this;
    }

    const entries = Array.from(this.store.entries());

    const fromIndex = entries.findIndex(byKey(key));
    const element = entries[fromIndex];

    if (!element) {
      return this;
    }

    entries.splice(fromIndex, 1);

    const relativeIndex = entries.findIndex(byKey(relativeKey));
    const toIndex = relativePosition(relativeIndex, order);

    entries.splice(toIndex, 0, element);

    this.store = new Map(entries);

    return this;
  };
}
