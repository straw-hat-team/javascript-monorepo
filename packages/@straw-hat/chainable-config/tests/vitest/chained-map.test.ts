import { expect, describe, test } from 'vitest';
import { ChainedMap } from '../../src/chained-map';
import { ChainedSet } from '../../src/chained-set';

test('validates if a value is a chainable map object', () => {
  const chainMap = new ChainedMap(undefined);
  expect(ChainedMap.isChainedMap(chainMap)).toBeTruthy();
  expect(ChainedMap.isChainedMap('helloi')).toBeFalsy();
});

describe('Given a ChainedMap object', () => {
  describe('When calling .set()', () => {
    test('Then saves the value', () => {
      const chaiMap = new ChainedMap(undefined);
      const context = chaiMap.set('name', 'pepega');
      expect(chaiMap.toConfig()).toEqual({ name: 'pepega' });
      expect(context).toBe(chaiMap);
    });
  });

  describe('When calling .delete()', () => {
    test('Then removes the value', () => {
      const chaiMap = new ChainedMap(undefined).merge({ name: 'pepega', reaction: 'pogchamp' });
      const context = chaiMap.delete('name');
      expect(chaiMap.toConfig()).toEqual({
        reaction: 'pogchamp',
      });
      expect(context).toBe(chaiMap);
    });
  });

  describe('When calling .keys()', () => {
    test('Then returns the list of keys', () => {
      const chaiMap = new ChainedMap(undefined).merge({ name: 'pepega', reaction: 'pogchamp' });
      const keys = Array.from(chaiMap.keys());
      expect(keys).toEqual(['name', 'reaction']);
    });
  });

  describe('When calling .clear()', () => {
    test('Then removes all the values', () => {
      const chaiMap = new ChainedMap(undefined).merge({ name: 'pepega', reaction: 'pogchamp' });
      const context = chaiMap.clear();
      expect(chaiMap.toConfig()).toEqual({});
      expect(context).toBe(chaiMap);
    });
  });

  describe('When calling .values()', () => {
    test('Then returns the list of values', () => {
      const chaiMap = new ChainedMap(undefined).merge({
        name: 'pepega',
        reaction: 'pogchamp',
        channel: 'alchemist_ubi',
      });
      const values = Array.from(chaiMap.values());
      expect(values).toEqual(['pepega', 'pogchamp', 'alchemist_ubi']);
    });
  });

  describe('When calling .entries()', () => {
    describe('When have entries', () => {
      test('Then returns the list of entries', () => {
        const chaiMap = new ChainedMap(undefined).merge({
          name: 'pepega',
          reaction: 'pogchamp',
          channel: 'alchemist_ubi',
        });
        const entries = Array.from(chaiMap.entries());
        expect(entries).toEqual([
          ['name', 'pepega'],
          ['reaction', 'pogchamp'],
          ['channel', 'alchemist_ubi'],
        ]);
      });
    });
  });

  describe('When calling .has()', () => {
    describe('When the value does not exists', () => {
      test('Then returns false', () => {
        const chaiMap = new ChainedMap(undefined);
        expect(chaiMap.has('name')).toBeFalsy();
      });
    });
    describe('When the value does exists', () => {
      test('Then returns true', () => {
        const chaiMap = new ChainedMap(undefined).set('name', 'pepega');
        expect(chaiMap.has('name')).toBeTruthy();
      });
    });
  });

  describe('When calling .merge()', () => {
    test('Then merges the values', () => {
      const chaiMap = new ChainedMap(undefined).merge({
        name: 'pepeg',
        hello: 'world',
        items: [1, 2, 3],
        complex: new ChainedMap(undefined).set('something', 'else'),
      });
      const context = chaiMap.merge({
        newKey: true,
        name: 'pepega',
        items: [4, 5, 6],
        complex: {
          more: 'stuff',
        },
      });
      const config = chaiMap.toConfig();
      expect(config).toEqual({
        complex: { something: 'else', more: 'stuff' },
        hello: 'world',
        items: [1, 2, 3, 4, 5, 6],
        name: 'pepega',
        newKey: true,
      });
      expect(context).toBe(chaiMap);
    });

    describe('When omitting keys', () => {
      test('Then merges the values omitting the keys', () => {
        const config = new ChainedMap(undefined)
          .merge(
            {
              name: 'pepeg',
              hello: 'world',
              items: [1, 2, 3],
            },
            ['items'],
          )
          .toConfig();
        expect(config).toEqual({ hello: 'world', name: 'pepeg' });
      });
    });
  });

  describe('When calling .getOrCompute()', () => {
    describe('When the key is not present', () => {
      test('Then sets and returns the value', () => {
        const chaiMap = new ChainedMap(undefined);
        expect(chaiMap.getOrCompute('a', () => 'alpha')).toBe('alpha');
        expect(chaiMap.toConfig()).toEqual({
          a: 'alpha',
        });
      });
    });

    describe('When the key is present', () => {
      test('Then returns the value', () => {
        const chaiMap = new ChainedMap(undefined).set('a', 1);
        expect(chaiMap.getOrCompute('a', () => 'alpha')).toBe(1);
        expect(chaiMap.toConfig()).toEqual({
          a: 1,
        });
      });
    });
  });

  describe('When calling .toConfig()', () => {
    describe('When does not have entries', () => {
      test('Then returns a empty configuration object', () => {
        const config = new ChainedMap(undefined).toConfig();
        expect(config).toEqual({});
      });
    });
    describe('When have some entries', () => {
      test('Then returns a configuration object', () => {
        const config = new ChainedMap(undefined)
          .set('^image![a-zA-Z0-9$_-]+$', 'GlobalImageStub', { alias: 'images' })
          .merge({
            name: 'pepeg',
            hello: 'world',
            answer: new ChainedMap(undefined).merge({
              name: 'pogchamp',
              hello: 'ubi',
            }),
          })
          .toConfig();
        expect(config).toEqual({
          name: 'pepeg',
          hello: 'world',
          '^image![a-zA-Z0-9$_-]+$': 'GlobalImageStub',
          answer: {
            name: 'pogchamp',
            hello: 'ubi',
          },
        });
      });
    });
  });

  describe('When asArray configuration is set', () => {
    test('Then returns the config as an array', () => {
      const config = new ChainedMap(undefined, { asArray: true })
        .set('js', 'js')
        .set('jsx', 'jsx')
        .set('complex', 'jsx')
        .set('alias', 'pepeg', { alias: 'myalias' })
        .set('complex', new ChainedSet(undefined).add('nice'))
        .toConfig();
      expect(config).toEqual(['js', 'jsx', ['nice'], 'pepeg']);
    });
  });

  describe('When emptyAsUndefined configuration is set', () => {
    test('Then returns the config as undefined', () => {
      const config = new ChainedMap(undefined, { asArray: true, emptyAsUndefined: true }).toConfig();
      expect(config).toEqual(undefined);
    });
  });

  describe('When calling .toString()', () => {
    test('Then returns the correct string', () => {
      const pepegMap = new ChainedMap(undefined, { name: 'config.pepeg' })
        .set('twitter', 'alchemist_ubi')
        .set('twitch', 'alchemist_ubi')
        .set('^image![a-zA-Z0-9$_-]+$', 'GlobalImageStub', { alias: 'images' });
      const complexSet = new ChainedSet(undefined).add('nice');
      const output = new ChainedMap(undefined, { asArray: true, name: 'config' })
        .set('js', 'js')
        .set('jsx', 'jsx')
        .set('complex', 'jsx')
        .set('complex', complexSet)
        .set('pepeg', pepegMap)
        .toString();

      expect(output).toMatchSnapshot();
    });
  });
});
