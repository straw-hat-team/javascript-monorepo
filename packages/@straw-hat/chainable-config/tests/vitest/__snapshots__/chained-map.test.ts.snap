// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given a ChainedMap object > When calling .toString() > Then returns the correct string 1`] = `"[/* config.get('js') */ 'js', /* config.get('jsx') */ 'jsx', /* config.get('complex') */ ['nice'], /* config.get('pepeg') */ {/* config.pepeg.get('twitter') */ 'twitter': 'alchemist_ubi', /* config.pepeg.get('twitch') */ 'twitch': 'alchemist_ubi', /* config.pepeg.get('images') */ '^image![a-zA-Z0-9$_-]+$': 'GlobalImageStub'}]"`;
